### PR TITLE
Add ability to edit a degree

### PR DESCRIPTION
**Pull request by chubberlisk**
### Context

Candidates need to be able to edit any degrees they add, currently when you click `Change` nothing happens as the `href=#`. This is the last part of allowing candidates to add their degrees.

### Changes proposed in this pull request

This PR allows candidates to edit a degree.

### Screenshots

image

image

image

### Guidance to review

@alexdesi I've done the renames here: https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/commit/e2ed0e000ecc6f9f236c41611b33be290e99497e and https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/commit/703024e8e324e7e42bc5d511e7a22aeda5073f6e. Check out the commit explanation and let me know what you think. I might have made it worse 😅 .

Would recommend going through commit by commit and trying it out.

### Link to Trello card

[194 - Adding degree information](https://trello.com/c/ADqlESXx/194-adding-degree-information)

### CODE REVIEW

**Comment by duncanjbrown** [comment]
If I enter a degree with a 2:1, save it, and click change on the review screen, I see `upper_second` in the "Other" box:

Screenshot 2019-11-05 at 16 51 22

**Comment by chubberlisk** [comment]
> If I enter a degree with a 2:1, save it, and click change on the review screen, I see `upper_second` in the "Other" box:
> 
> Screenshot 2019-11-05 at 16 51 22

Nice spot! Was due to some refactoring earlier. Fixed in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/441/commits/3921104861fadf86c416a37e904284f666f02a7c.

**Comment by duncanjbrown** [comment]
This is pretty good! Two questions:

- one about the duplication between `save` and `update` in the `DegreesForm`, inline
- it feels a bit funny to me that `build_all_from_application` values differ from `build_from_application`. Why is that?

And one small ?bug: the "back" link when editing a degree goes back to the application page, not the section review page as it does for Work History

**Comment by chubberlisk** [comment]
> This is pretty good! Two questions:
> 
> * one about the duplication between `save` and `update` in the `DegreesForm`, inline

What do you believe is duplicated?

> * it feels a bit funny to me that `build_all_from_application` values differ from `build_from_application`. Why is that?

Ah! They should be returning the same values when creating a new `DegreesForm`! I will update this and refactor both into a private method.

> And one small ?bug: the "back" link when editing a degree goes back to the application page, not the section review page as it does for Work History

Will fix this!

**Comment by duncanjbrown** [comment]
Oops — forgot to submit my review comments 🙄. Duplication detail there

File: app/controllers/candidate_interface/degrees/base_controller.rb [hunk 1]
```diff
@@ -10,7 +10,23 @@ def create
...
-      if @degree.save_base(application_form)
+      if @degree.save(application_form)
+        redirect_to candidate_interface_degrees_review_path
+      else
+        render_new
+      end
+    end
+
+    def edit
+      application_from = current_candidate.current_application
```
[the remark this answers is not in the document]
Comment (chubberlisk) [reply app/controllers/candidate_interface/degrees/base_controller.rb]: 👀 I done this before, I can't seem to type application form 🤦‍♀

File: app/models/candidate_interface/degrees_form.rb [hunk 2]
```diff
@@ -80,6 +80,21 @@ def save_base(application_form)
       true
     end
 
+    def update(application_form, degree_id)
```
[the remark this answers is not in the document]
Comment (chubberlisk) [reply app/models/candidate_interface/degrees_form.rb]: We paired on this. Decided to get rid of `degree_id` by using a hidden field. Updated in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/441/commits/2842deb3dda86f4739d96c6f33af71995dac028a.

**Comment by chubberlisk** [comment]
@duncanjbrown I believe I have addressed all your comments here and as we discussed when pairing.

Highly recommend going through the commits as it's made the diff very noisy due to the renaming of things, sorry!

File: app/controllers/candidate_interface/degrees/destroy_controller.rb [hunk 3]
```diff
@@ -2,22 +2,22 @@ module CandidateInterface
...
         .destroy!
 
       redirect_to candidate_interface_degrees_review_path
     end
 
   private
 
-    def degrees_params
-      params.permit(:id)
+    def degree_id_params
```
Comment (duncanjbrown) [inline app/controllers/candidate_interface/degrees/destroy_controller.rb]: `params` is misleading if it returns a single value — maybe `current_degree_id`?

**Comment by duncanjbrown** [comment]
:+1:

**Pull request merged**